python_modules/dagster-graphql/dagster_graphql/implementation/fetch_runs.py

```diff
@@ -379,3 +379,59 @@ def get_logs_for_run(
         cursor=conn.cursor,
         hasMore=conn.has_more,
     )
+
+
+def _fetch_runs_not_in_backfill(
+    instance, cursor: Optional[str], limit: Optional[int]
+) -> Sequence[RunRecord]:
+    if limit is None:
+        new_runs = instance.get_run_records(cursor=cursor)
+        return [run for run in new_runs if run.dagster_run.tags.get(BACKFILL_ID_TAG) is None]
+
+    runs = []
+    while len(runs) < limit:
+        new_runs = instance.get_run_records(limit=limit, cursor=cursor)
+        if len(new_runs) == 0:
+            return runs
+        cursor = new_runs[-1].dagster_run.run_id
+        runs.extend([run for run in new_runs if run.dagster_run.tags.get(BACKFILL_ID_TAG) is None])
+
+    return runs[:limit]
+
+
+def get_mega_runs(
+    graphene_info: "ResolveInfo",
+    cursor: Optional[str] = None,
+    limit: Optional[int] = None,
+) -> Sequence["GrapheneMegaRun"]:
+    """Returns a merged list of backfills and single runs (runs that are not part of a backfill).
+
+    Cursor format: run_id;backfill_id
+    """
+    from ..schema.pipelines.pipeline import GrapheneMegaRun
+
+    check.opt_str_param(cursor, "cursor")
+    check.opt_int_param(limit, "limit")
+
+    instance = graphene_info.context.instance
+
+    run_cursor, backfill_cursor = cursor.split(";") if cursor else (None, None)
```

Comment:
cursor format: For querying runs, the cursor is the ID of the last run in the returned list. I started with something similar here where the cursor is the last run id and the last backfill id separated by a ";". The difficulty with that is that users would need to iterate through the returned list to find the last run/backfill. We could add a resolver on `GrapheneMegaRuns` that returns the cursor for them (i think that's reasonable), but i also want to think about cursor formats that might be easier to generate in the first place.

Reply:
heres what I did for similar problem recently https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/remote_representation/external.py#L85-L112

the cursor should be opaque to the client, it just gets echoed back to the server to fetch more